airflow/jobs/local_task_job.py

```diff
@@ -208,6 +208,14 @@ def heartbeat_callback(self, session=None):
                 )
                 raise AirflowException("PID of job runner does not match")
         elif self.task_runner.return_code() is None and hasattr(self.task_runner, 'process'):
+            if ti.state == State.SKIPPED:
+                # if the task was externally marked as skipped, its likely due to a timeout.
+                dagrun = ti.get_dagrun()
+                if dagrun.end_date:
+                    execution_time = dagrun.end_date - dagrun.start_date
+                    dagrun_timeout = dagrun.get_dag().dagrun_timeout
+                    if dagrun_timeout and execution_time > dagrun_timeout:
+                        self.log.warning("DagRun timed out after %s.", str(execution_time))
```

Comment:
```suggestion
                dagrun = ti.dag_run
                if dagrun.end_date:
                    execution_time = dagrun.end_date - dagrun.start_date
                    dagrun_timeout = ti.task.dag.dagrun_timeout
                    if dagrun_timeout and execution_time > dagrun_timeout:
                        self.log.warning("DagRun timed out after %s.", str(execution_time))
```
Can you try if this will work

Reply:
Yeah, I'll try this today and let you know if it works.

Reply:
This appears to work so I've updated this PR.

Reply:
I think we should use `get_dagrun()` because it does caching better than `dag_run`? Not sure if it makes a difference here.

Reply:
It looks like `ti.get_dagrun` will use the cached value of `ti.dag_run` if it exists, otherwise it will fall back to querying the database. It's probably the safest option to use here in case for whatever reason `ti.dag_run` doesn't exist.

Reply:
agree